File: templates/production-adapters/opencode-http-session/python/runner_inputs.py
from __future__ import annotations

import json
import re
import shutil
from pathlib import Path
from typing import Any


class RunnerInputError(RuntimeError):
    pass


def resolve_run_dir(
    *,
    repo_root: Path,
    run_root: Path,
    run_dir_arg: str | None,
    derive_from_input: str,
    workflow_root: Path,
    input_artifact: str,
) -> Path:
    if run_dir_arg is not None:
        return Path(run_dir_arg).resolve()
    payload = read_json_object(workflow_root / input_artifact)
    fields = [field.strip() for field in derive_from_input.split("+") if field.strip()]
    if not fields:
        raise RunnerInputError("runtime.runDirectory.deriveFromInput must name at least one input field")
    slug = "--".join(sanitize_segment(_required_string(payload, field)) for field in fields)
    return (run_root / slug).resolve()


def prepare_run_directory(
    *,
    run_dir: Path,
    run_root: Path,
    workflow_root: Path,
    seed_paths: tuple[Path, ...],
    marker_name: str,
    fresh_run: bool,
) -> None:
    if fresh_run and run_dir.exists():
        ensure_safe_fresh_run(run_dir=run_dir, run_root=run_root, marker_name=marker_name)
        shutil.rmtree(run_dir)
    for relative_path in seed_paths:
        source_path = workflow_root / relative_path
        target_path = run_dir / relative_path
        if not source_path.exists():
            raise RunnerInputError(f"seed artifact does not exist: {source_path}")
        if not target_path.exists():
            target_path.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(source_path, target_path)
    write_marker(run_dir, marker_name=marker_name)


def ensure_safe_fresh_run(*, run_dir: Path, run_root: Path, marker_name: str) -> None:
    marker_path = run_dir / marker_name
    try:
        run_dir.relative_to(run_root.resolve())
        return
    except ValueError:
        pass
    if marker_path.is_file():
        return
    raise RunnerInputError(f"refusing --fresh-run outside canonical run root without runner marker: {run_dir}")


def write_marker(run_dir: Path, *, marker_name: str) -> None:
    marker_path = run_dir / marker_name
    marker_path.parent.mkdir(parents=True, exist_ok=True)
    marker_path.write_text("opencode-http-session\n", encoding="utf-8")


def read_json_object(path: Path) -> dict[str, Any]:
    payload = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise RunnerInputError(f"expected JSON object: {path}")
    return dict(payload)


def sanitize_segment(value: str) -> str:
    normalized = re.sub(r"[^A-Za-z0-9._-]+", "-", value.strip())
    return normalized.strip("-._") or "run"


def _required_string(payload: dict[str, Any], key: str) -> str:
    value = payload.get(key)
    if not isinstance(value, str) or not value.strip():
        raise RunnerInputError(f"input artifact is missing non-empty string field: {key}")
    return value

File: templates/production-adapters/opencode-http-session/typescript/runner-inputs.ts
import { copyFile, rm } from "node:fs/promises"
import { dirname, join, resolve } from "node:path"

import { readJsonObject, writeText } from "./json-store"

export async function resolveRunDirectory(input: {
  readonly deriveFromInput: string
  readonly explicitRunDirectory?: string
  readonly inputArtifact: string
  readonly repoRoot: string
  readonly runRoot: string
  readonly workflowRoot: string
}): Promise<string> {
  if (input.explicitRunDirectory !== undefined) {
    return resolve(input.explicitRunDirectory)
  }
  const payload = await readJsonObject(join(input.workflowRoot, input.inputArtifact))
  const fields = input.deriveFromInput.split("+").map((field) => field.trim()).filter((field) => field.length > 0)
  if (fields.length === 0) {
    throw new Error("runtime.runDirectory.deriveFromInput must name at least one input field")
  }
  const slug = fields.map((field) => sanitizeSegment(readRequiredString(payload, field))).join("--")
  return resolve(input.runRoot, slug)
}

export async function prepareRunDirectory(input: {
  readonly freshRun: boolean
  readonly markerName: string
  readonly runDirectory: string
  readonly runRoot: string
  readonly seedArtifacts: readonly string[]
  readonly workflowRoot: string
}): Promise<void> {
  if (input.freshRun && (await Bun.file(input.runDirectory).exists())) {
    await ensureSafeFreshRun(input.runDirectory, input.runRoot, input.markerName)
    await rm(input.runDirectory, { force: true, recursive: true })
  }
  for (const relativePath of input.seedArtifacts) {
    const source = join(input.workflowRoot, relativePath)
    const target = join(input.runDirectory, relativePath)
    if (!(await Bun.file(source).exists())) {
      throw new Error(`seed artifact does not exist: ${source}`)
    }
    await Bun.$`mkdir -p ${dirname(target)}`.quiet()
    await copyFile(source, target)
  }
  await writeText(join(input.runDirectory, input.markerName), "opencode-http-session\n")
}

async function ensureSafeFreshRun(runDirectory: string, runRoot: string, markerName: string): Promise<void> {
  const resolvedRunDirectory = resolve(runDirectory)
  const resolvedRunRoot = resolve(runRoot)
  const insideRunRoot = resolvedRunDirectory === resolvedRunRoot || resolvedRunDirectory.startsWith(`${resolvedRunRoot}/`)
  if (insideRunRoot || (await Bun.file(join(runDirectory, markerName)).exists())) {
    return
  }
  throw new Error(`refusing --fresh-run outside canonical run root without runner marker: ${runDirectory}`)
}

function readRequiredString(payload: Record<string, unknown>, key: string): string {
  const value = payload[key]
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new Error(`input artifact is missing non-empty string field: ${key}`)
  }
  return value
}

function sanitizeSegment(value: string): string {
  return value.replaceAll(/[^A-Za-z0-9._-]+/g, "-").replaceAll(/^[._-]+|[._-]+$/g, "") || "run"
}
